data/transactions/logic/eval.go

```diff
@@ -1350,23 +1397,21 @@ func (cx *EvalContext) step() error {
 			return fmt.Errorf("%3d %s returned 0 cost", cx.pc, spec.Name)
 		}
 	}
-	cx.cost += opcost
-	if cx.PooledApplicationBudget != nil {
-		*cx.PooledApplicationBudget -= opcost
-	}
 
-	if cx.remainingBudget() < 0 {
-		// We're not going to execute the instruction, so give the cost back.
-		// This only matters if this is an inner ClearState - the caller should
-		// not be over debited. (Normally, failure causes total txtree failure.)
-		cx.cost -= opcost
-		if cx.PooledApplicationBudget != nil {
-			*cx.PooledApplicationBudget += opcost
-		}
+	if opcost > cx.remainingBudget() {
```

Comment:
this seems more restrictive than before?

Reply:
It's not intended to be. Let me make the case.

Previously, we calculated the cost of the next opcode `opcost` and then performed the cost increment / budget decrement. Then we asked, "would that be going too far?" (`cx.remainingBudget() < 0`). If so, we undid the cost increment / budget decrement.

Now, we calculate that next cost `opcost` and check beforehand `opcost > cx.remainingBudget()` and fail if it is. This way, there's no need to back out a change that was illegal in the first place.

But the intent is to fail under the exact same conditions.

Reply:
i think I missed the relocation of `cx.cost +=`